### PR TITLE
Fixed executing editor.destroy() during image upload

### LINKED ISSUE #966: Executing editor.destroy() during image upload throws error
## Are you reporting a feature request or a bug?
Bug 

## Provide detailed reproduction steps (if any)

1. Start upload large file
2. Destroy editor

### Expected result

*What is the expected result of the above steps?*
Uploading are canceled.

### Actual result
Editor are crashed with error.

*What is the actual result of the above steps?*

## Other details

* Browser: All
* OS: Window
* CKEditor version: 4.7.1
* Installed CKEditor plugins: default plugins

### PR DESCRIPTION
**Pull request by jacekbogdanski**
## What is the purpose of this pull request?

Bug fix.

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

Handled [`destroyed`](https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#destroy) editor status during [file upload](https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_fileTools_uploadWidgetDefinition.html#onUploading).

Closes #966

### CODE REVIEW

File: tests/plugins/uploadwidget/manual/destroyeditor.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<div id="editor"></div>
+<button id="destroyer">Destroy</button>
+
+<script>
+	CKEDITOR.replace( 'editor', {
```
Comment (engineering-this) [inline tests/plugins/uploadwidget/manual/destroyeditor.html]: As this test requires drag and drop it should be ignored on mobile devices.

**Comment by Comandeer** [comment]
Rebased onto latest `major`.

File: plugins/uploadwidget/plugin.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿/**
+/**
```
Comment (Comandeer) [inline plugins/uploadwidget/plugin.js]: What exactly was changed here? 🤔

File: tests/plugins/uploadwidget/manual/destroyeditor.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+		imageUploadUrl: 'http://sub.ckeditor.dev/',
+		on: {
+				instanceReady: function( evt ) {
+					var editor = evt.editor;
+
+					document.getElementById( 'destroyer' ).addEventListener( 'click', function() {
```
Comment (Comandeer) [inline tests/plugins/uploadwidget/manual/destroyeditor.html]: IE 8 does not support `addEventListener`, which results in error in this browser and not ignoring the test.

File: tests/plugins/uploadwidget/manual/destroyeditor.html [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+						editor.destroy();
+					} );
+
+					if ( !CKEDITOR.fileTools.isFileUploadSupported || bender.tools.env.mobile ) {
```
Comment (Comandeer) [inline tests/plugins/uploadwidget/manual/destroyeditor.html]: This should be moved to the beginning and return early with ignoring. Such move would prevent mentioned errors in IE 8.

File: tests/plugins/uploadwidget/manual/destroyeditor.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+@bender-tags: 4.9.1, 966, bug
```
Comment (Comandeer) [inline tests/plugins/uploadwidget/manual/destroyeditor.md]: Please update version tag.

File: tests/plugins/uploadwidget/uploadwidget.js [hunk 6]
```diff
@@ -156,6 +156,26 @@
...
+			bender.editorBot.create( { name: 'destroy_test', config: { extraPlugins: 'uploadwidget' } }, function( bot ) {
+				var editor = bot.editor;
+
+				addTestUploadWidget( editor, 'testuploadwidget' );
+				pasteFiles( editor, [ bender.tools.getTestPngFile() ] );
+
+				var loader = editor.uploadRepository.loaders[ 0 ];
+
+				try {
+					editor.destroy();
```
Comment (Comandeer) [inline tests/plugins/uploadwidget/uploadwidget.js]: I'm not sure if it's really "destroying while uploading". In reality dropping image generates three states:

1. `loading`
2. `loaded`
3. `uploading`

Destroying editor via button in manual test is in fact possible only **after** switching to `uploading`. What's more, error is thrown not while changing state, but while handling `update` event → https://github.com/ckeditor/ckeditor-dev/blob/f4e106d5736e5d4211f7371a863d17dd46270e30/plugins/uploadwidget/plugin.js#L257

File: plugins/uploadwidget/plugin.js [hunk 7]
```diff
@@ -252,6 +252,11 @@
 					oldStyle, newStyle;
 
 				loader.on( 'update', function( evt ) {
+					// Abort if the editor has been destroyed.
+					if ( widget.editor.status === 'destroyed' ) {
+						return;
```
Comment (Comandeer) [inline plugins/uploadwidget/plugin.js]: `return` does not prevent the continuation of file upload, it still happens in the background. If editor is destroyed, loader should be aborted (`loader.abort()`).

File: plugins/uploadwidget/plugin.js [hunk 8]
```diff
@@ -252,6 +252,11 @@
...
+					}
+
 					// Abort if widget was removed.
 					if ( !widget.wrapper || !widget.wrapper.getParent() ) {
 						if ( !editor.editable().find( '[data-cke-upload-id="' + id + '"]' ).count() ) {
```
Comment (Comandeer) [inline plugins/uploadwidget/plugin.js]: In fact it should be enough for fixing this issue to enhance this conditional to check if `editor.editable` exists.

File: tests/plugins/uploadwidget/uploadwidget.js [hunk 9]
```diff
@@ -156,6 +156,26 @@
 			} );
 		},
 
+		'test destroy editor during upload': function() {
```
Comment (Comandeer) [inline tests/plugins/uploadwidget/uploadwidget.js]: Please add issue reference.

File: plugins/uploadwidget/plugin.js (hunk 2 above)
Comment (jacekbogdanski) [reply plugins/uploadwidget/plugin.js]: BOM removed.

**Comment by mlewand** [comment]
Funnily enough I did not notice this PR when I started to review #967. Issue is fixed and the tests are added so I'm closing this PR too.

**Pull request closed without merging**